subcommand/consul-sidecar/command.go

```diff
@@ -80,57 +96,165 @@ func (c *Command) Run(args []string) int {
+		server = c.createMergedMetricsServer()
+
+		// Run the merged metrics server
+		c.logger.Info("Running merged metrics server.")
+		go func() {
+			if err = server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+				c.logger.Error(err.Error())
+			}
+		}()
+	}
+
+	// The work loop for re-registering the service. We continually re-register
+	// our service every syncPeriod. Consul is smart enough to know when the
+	// service hasn't changed and so won't update any indices. This means we
+	// won't be causing a lot of traffic within the cluster. We tolerate Consul
+	// Clients going down and will simply re-register once it's back up.
 	//
 	// The loop will only exit when the Pod is shut down and we receive a SIGINT.
 	for {
 		start := time.Now()
-		cmd := exec.CommandContext(ctx, c.flagConsulBinary, c.consulCommand...)
+		cmd := exec.CommandContext(signalCtx, c.flagConsulBinary, c.consulCommand...)
 
 		// Run the command and record the stdout and stderr output
 		output, err := cmd.CombinedOutput()
 		if err != nil {
-			logger.Error("failed to sync service", "output", strings.TrimSpace(string(output)), "err", err, "duration", time.Since(start))
+			c.logger.Error("failed to sync service", "output", strings.TrimSpace(string(output)), "err", err, "duration", time.Since(start))
 		} else {
-			logger.Info("successfully synced service", "output", strings.TrimSpace(string(output)), "duration", time.Since(start))
+			c.logger.Info("successfully synced service", "output", strings.TrimSpace(string(output)), "duration", time.Since(start))
 		}
 		select {
 		// Re-loop after syncPeriod or exit if we receive interrupt or terminate signals.
 		case <-time.After(c.flagSyncPeriod):
 			continue
-		case <-ctx.Done():
+		case <-signalCtx.Done():
+			// After the signal is received, wait for the merged metrics server
+			// to gracefully shutdown as well if it has been enabled. This can
+			// take up to metricsServerShutdownTimeout seconds.
+			if c.flagEnableMetricsMerging {
+				c.logger.Info("Attempting to shut down metrics server.")
+				c.shutdownMetricsServer(server)
+			}
 			return 0
 		}
 	}
 }
 
+// shutdownMetricsServer handles gracefully shutting down the server. This will
+// call server.Shutdown(), which will indefinitely wait for connections to turn
+// idle. To avoid potentially waiting forever, we pass a context to
+// server.Shutdown() that will timeout in metricsServerShutdownTimeout (5) seconds.
+func (c *Command) shutdownMetricsServer(server *http.Server) {
+	shutdownCtx, shutdownCancelFunc := context.WithTimeout(context.Background(), metricsServerShutdownTimeout)
+	defer shutdownCancelFunc()
+
+	// If the server hasn't started up yet before this function is called, it
+	// could be nil.
+	if server != nil {
+		c.logger.Info("Merged metrics server exists, attempting to gracefully shut down server.")
+		if err := server.Shutdown(shutdownCtx); err != nil {
+			c.logger.Error(fmt.Sprintf("Server shutdown failed: %s", err))
```

Comment:
Should we return/exit from here so we don't print the next line in failure cases?

Reply:
Nice catch!